packages/schema-derive/src/lib.rs

```diff
@@ -0,0 +1,89 @@
+use proc_macro2::TokenStream;
+use quote::quote;
+use syn::{parse_macro_input, ItemEnum, Type, Variant};
+
+/// Extract the query -> response mapping out of an enum variant.
+fn parse_query(v: Variant) -> TokenStream {
+    let query = to_snake_case(&v.ident.to_string());
```

Comment:
Our derive macro could panic if the `rename_all = "snake_case"` attribute is missing, I suppose. But we'd have to write (or borrow) code to parse them the same way serde parses them. I don't know if I like that.

Reply:
Panic sounds good. The other option is to let `QueryResponses` do the serde configuration. Then users can remove the line and get the right (tm) thing from us.

Reply:
So the problem is a derive macro cannot rewrite the enum or add attributes to it. It can only add extra code (a trait implementation usually).

We'd need to use an attribute macro and that feels more heavy-handed. Attribute macros consume the input and allow us to output a modified version of it.

Reply:
Okay. But that means we can add extra code as part of `response_schemas()` that does a bit of serialization and check if snake case was used?

Reply:
Huh. Maybe you're onto something. I think if we add a trait bound like this:

```rust
trait QueryResponses: JsonSchema {
    // ...
}
```

then `response_schemas()` should be able to inspect the schema with something like this?

```rust
let schema = schema_for!(Self);
```

And maybe checking the schema for the presence of all the expected variants is a better integrity check than checking if a serde attribute is present? It sounds pretty good.

By the way, the parsing of serde attributes might not actually be hard. [Serde exposes some internals for that.](https://github.com/serde-rs/serde/tree/master/serde_derive_internals)

Reply:
Okay, so.
- The derive macro checking for serde attributes is a no go - it can be done in theory, but requires `serde_derive_internals`, which then causes dependency resolution problems in contracts (another dep tries to pull an earlier version of it). It's probably best to stay away from that can of worms and just not risk dependency problems for the future there.
- I'm going for the runtime integrity check described above since it just seems like a damn good idea to have. That throws an error during schema generation though, so pretty late.
- If we add [this](https://github.com/CosmWasm/cosmwasm/issues/1307#issuecomment-1171943798), we're golden.

Reply:
```bash
> cargo schema
--- snip ---
thread 'main' panicked at '{}: InconsistentQueries { query_msg: ["get-int", "other-balance", "recurse", "verifier"], responses: ["get_int", "other_balance", "recurse", "verifier"] }', examples/schema.rs:41:49
note: run with `RUST_BACKTRACE=1` environment variable to display a backtrace
```